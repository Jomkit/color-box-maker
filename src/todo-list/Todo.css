button {
    background: none;
    border: none;
    cursor: pointer;
}

.Todo-completeBtn {
    text-decoration: underline;
    color: hsl(0, 0%, 45%);
}

.Todo-completeBtn:hover {
    color: hsl(0, 0%, 28%);

}

.Todo-editBtn {
    text-decoration: underline;
    margin: 5px;
    padding: 3px;
}

.Todo-editBtn:hover {
    background: lightblue;
    border-radius: 3px;
}

.Todo-deleteBtn {
    color: red;
}
