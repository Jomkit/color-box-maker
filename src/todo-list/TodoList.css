.TodoList {
    display: float;
    background: gainsboro;
    border-radius: 10px;
    margin: 20%;
    margin-top: 10%;
    padding-bottom: 20px;
}

.TodoList-divider {
    margin: 10px 30px;
}