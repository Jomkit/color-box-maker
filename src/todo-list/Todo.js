import React, { useState } from 'react'
import "./Todo.css";
import Form from './Form';

const Todo = ({ id, task, completed, deleteTodo, editTodo, completeTodo }) => {
    const [formOpen, setFormOpen] = useState(false);
    const toggleEdit = () => {setFormOpen(!formOpen)}
  return (
    <>
        <li style={{textDecoration: completed && "line-through"}}>
            Task: {task} 
            <button className='Todo-completeBtn' onClick={completeTodo}>Mark as completed</button>
            <button className='Todo-editBtn' onClick={toggleEdit} >Edit</button>
            <button className='Todo-deleteBtn' onClick={deleteTodo}><span>&#9746;</span></button>
        </li>
        {formOpen && <Form id={id} toggleEdit={toggleEdit} editTodo={editTodo} />}
    </>
  )
}

export default Todo