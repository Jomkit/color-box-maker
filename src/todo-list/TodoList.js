import React, { useState } from 'react'
import { v4 as uuid } from "uuid";
import "./TodoList.css";

// Components
import Todo from './Todo';
import NewTodoForm from './NewTodoForm';

const TodoList = () => {
    const exampleTodos = [
        { id: uuid(), completed: true, task: "Walk dog" },
        { id: uuid(), completed: false, task: "Mow lawn" },
        { id: uuid(), completed: false, task: "Water flowers" },
    ]
    const [todos, setTodos] = useState(
        localStorage.getItem('todos') ? 
        JSON.parse(localStorage.getItem('todos')) :
        exampleTodos
    );

    // localStorage helper functions
    const saveToLocalStorage = (updatedTodos) => {
        const payload = JSON.stringify(updatedTodos);
        localStorage.setItem('todos', payload);
    }
    
    const addTodo = (task) => {
        const updatedTodos = [
            ...todos.map(todo => ({...todo})),
            {id: uuid(), completed: false, task: task}
        ];
        setTodos(updatedTodos);
        saveToLocalStorage(updatedTodos);
    };
    const deleteTodo = (id) => {
        const updatedTodos = [...todos.filter(todo => todo.id !== id)];
        setTodos(updatedTodos);
        saveToLocalStorage(updatedTodos);
    };
    const editTodo = (id, newTask) => {
        const updatedTodos = [
            ...todos.map(todo => 
                todo.id === id 
                ? {...todo, task: newTask} 
                : todo
            )];
        setTodos(updatedTodos);
        saveToLocalStorage(updatedTodos);
    };
    const completeTodo = (id) => {
        const updatedTodos = [
            ...todos.map(todo => 
                todo.id === id 
                ? {...todo, completed: !todo.completed} 
                : todo
            )];
        setTodos(updatedTodos);
        saveToLocalStorage(updatedTodos);
    };
    const resetTodos = () => {
        setTodos(exampleTodos);
        localStorage.removeItem("todos");
    }
    
  return (
    <div className='TodoList'>
        <h1>Todo</h1>
        <NewTodoForm addTodo={addTodo} />
        <button className='TodoList-resetBtn' onClick={resetTodos}>Reset Todos</button>
        <hr className='TodoList-divider'></hr>
        <ul className='TodoList-ul'>
            {todos.map(({ id, task, completed }) => (
                <Todo 
                key={id}
                 id={id} 
                 task={task} 
                 completed={completed} 
                 deleteTodo={() => deleteTodo(id)} 
                 editTodo={editTodo} 
                 completeTodo={() => completeTodo(id)}
                />
            ))}
        </ul>
    </div>
  )
}

export default TodoList