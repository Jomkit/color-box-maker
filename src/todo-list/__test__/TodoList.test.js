import { render, fireEvent } from "@testing-library/react";
import TodoList from "../TodoList";
import Todo from "../Todo";

it('should render without crashing', () => { 
    render(<TodoList />);
})

it('should match snapshot', () => { 
    const { asFragment } = render(<TodoList />);
    expect(asFragment()).toMatchSnapshot();
})

it('should add new todo', () => { 
    const { queryByText, getByLabelText } = render(<TodoList />);

    // No new todo
    expect(queryByText("Test add todo", {exact: false})).not.toBeInTheDocument();
    
    const taskInput = getByLabelText("Description:");
    const addTaskBtn = queryByText("Add Task");
    
    //Fill out the form
    fireEvent.change(taskInput, {target: {value: "Test add todo"}});
    fireEvent.click(addTaskBtn);
    
    expect(queryByText("Test add todo", {exact: false})).toBeInTheDocument();
 })

it('should delete a todo item', () => { 
    const { queryByText, debug } = render(<TodoList />);

    //Check that item to delete exists
    const task = queryByText("Water flowers", {exact: false});
    expect(task).toBeInTheDocument();
    
    const deleteBtn = task.querySelector(".Todo-deleteBtn");
    expect(deleteBtn).toBeInTheDocument();
    fireEvent.click(deleteBtn);
    expect(task).not.toBeInTheDocument();
})

it('should strikethrough an item when "Mark as complete" is pressed', () => { 
    const { queryByText } = render(<TodoList />);

    //Check that item to mark as complete exists
    const task = queryByText("Water flowers", {exact: false});
    expect(task).toBeInTheDocument();
    expect(task).not.toHaveStyle({textDecoration: "line-through"});

    const completeBtn = task.querySelector(".Todo-completeBtn");
    fireEvent.click(completeBtn);
    expect(task).toHaveStyle({textDecoration: "line-through"});
})

it('should edit a task', () => { 
    const { container, queryByText, queryByLabelText, debug } = render(<TodoList />);

    expect(container.querySelectorAll(".Todo-editBtn").length).toEqual(3);
    
    //check that the item to edit exists
    const task = queryByText("Water flowers", {exact: false});
    expect(task).toBeInTheDocument();
    
    //Check that edit task form appears
    expect(queryByLabelText("Edit Task:", {exact: false})).not.toBeInTheDocument();
    
    const toggleEditBtn = task.querySelector(".Todo-editBtn");
    fireEvent.click(toggleEditBtn);
    const editInput = queryByLabelText("Edit Task:", {exact: false});
    expect(editInput).toBeInTheDocument();
    
    const submitEditBtn = queryByText("Submit");
    expect(submitEditBtn).toBeInTheDocument();
    
    fireEvent.change(editInput, {target: {value: "Testing edit task"}});
    fireEvent.click(submitEditBtn);
    expect(queryByText("Testing edit task", {exact: false})).toBeInTheDocument();
    
    // number of tasks hasn't changed
    expect(container.querySelectorAll(".Todo-editBtn").length).toEqual(3);
})

afterEach(() => {
    localStorage.clear();
})

