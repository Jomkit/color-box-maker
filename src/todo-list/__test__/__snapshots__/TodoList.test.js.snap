// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="TodoList"
  >
    <h1>
      Todo
    </h1>
    <form
      class="NewTodoForm"
    >
      <label
        for="task"
      >
        Description: 
      </label>
      <input
        id="task"
        name="task"
        placeholder="Grab milk... etc."
        type="text"
        value=""
      />
      <button>
        Add Task
      </button>
    </form>
    <button
      class="TodoList-resetBtn"
    >
      Reset Todos
    </button>
    <hr
      class="TodoList-divider"
    />
    <ul
      class="TodoList-ul"
    >
      <li
        style="text-decoration: line-through;"
      >
        Task: Walk dog
        <button
          class="Todo-completeBtn"
        >
          Mark as completed
        </button>
        <button
          class="Todo-editBtn"
        >
          Edit
        </button>
        <button
          class="Todo-deleteBtn"
        >
          <span>
            ☒
          </span>
        </button>
      </li>
      <li>
        Task: Mow lawn
        <button
          class="Todo-completeBtn"
        >
          Mark as completed
        </button>
        <button
          class="Todo-editBtn"
        >
          Edit
        </button>
        <button
          class="Todo-deleteBtn"
        >
          <span>
            ☒
          </span>
        </button>
      </li>
      <li>
        Task: Water flowers
        <button
          class="Todo-completeBtn"
        >
          Mark as completed
        </button>
        <button
          class="Todo-editBtn"
        >
          Edit
        </button>
        <button
          class="Todo-deleteBtn"
        >
          <span>
            ☒
          </span>
        </button>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
